connectors/sources/onedrive.py

```diff
@@ -637,24 +771,22 @@ async def get_docs(self, filtering=None):
                     async for entity, download_url in self.client.get_owned_files(
                         user_id, skipped_extensions, pattern
                     ):
-                        entity = self.prepare_doc(entity)
-                        if entity["type"] == FILE and download_url:
-                            yield entity, partial(
-                                self.get_content, entity.copy(), download_url
-                            )
-                        else:
-                            yield entity, None
+                        yield self.send_document_to_es(entity, download_url)
         else:
+            requests = []
             async for user in self.client.list_users():
-                user_id = user.get("id")
-
-                async for entity, download_url in self.client.get_owned_files(user_id):
-                    entity = self.prepare_doc(entity)
-                    if entity["type"] == FILE and download_url:
-                        yield await self._decorate_with_access_control(
-                            entity, user_id, entity.get("_id")
-                        ), partial(self.get_content, entity.copy(), download_url)
-                    else:
-                        yield await self._decorate_with_access_control(
-                            entity, user_id, entity.get("_id")
-                        ), None
+                requests.append(self.build_owned_files_url(user))
+
+            async for response in self.json_batching(batched_apis=requests):
+                files = response.get("body", {}).get("value", [])
+                if entities := [file for file in files if file.get("name") != "root"]:
+                    if self._dls_enabled():
+                        entities = await self._bounbed_concurrent_tasks(
```

Comment:
2 concerns:

1. `concurrent_downloads` is meant to limit the number of concurrent downloads, not the call to list file permissions.
2. With `asyncio.gather`, it does improve the performance as we have multiple requests at the same time, but it doesn't reduce the number of calls to OneDrive.

I think this is another perfect use case of `json_batching`.

Reply:
Sadly, I did try json_batching here but it resulted into numerouse 429 errors. I then found json_batching didn't help to reduce rate-limit scenarios, because the number of resource count of batch request in Microsoft graph is counted as the sum of individual requests. So if 20 requests are batched, the resource count is 20 and not 1.